core/src/main/scala/reconciler/Reconciler.scala

```diff
@@ -119,6 +119,88 @@ abstract class Reconciler[S <: CustomResource[_, _], T <: ObjectResource] {
+  case class CacheEntryKey(name: String, namespace: String, kind: String)
+  case class Cache(events: List[Event], cache: Map[CacheEntryKey, CacheEntry])
+
+  def watchSourceRaw(implicit context: RequestContext,
+                     sourceFormat: Format[S],
+                     sourceListFormat: Format[ListResource[S]],
+                     sourceResourceDefinition: ResourceDefinition[S],
+                     sourceListResourceDefinition: ResourceDefinition[ListResource[S]],
+                     ec: ExecutionContext): Future[akka.stream.scaladsl.Source[WatchEvent[S], NotUsed]] = {
+    context.list[ListResource[S]].map { l =>
+      val initialSource = Source(l.items.map(l => WatchEvent(EventType.MODIFIED, l)))
+      val watchedSource = context.watchAllContinuously[S](Some(l.resourceVersion))
+
+      initialSource.concat(watchedSource)
+    }
+  }
+
+  def watchTargetRaw(implicit context: RequestContext,
+                     targetFormat: Format[T],
+                     targetListFormat: Format[ListResource[T]],
+                     targetResourceDefinition: ResourceDefinition[T],
+                     targetListResourceDefinition: ResourceDefinition[ListResource[T]],
+                     ec: ExecutionContext): Future[akka.stream.scaladsl.Source[WatchEvent[T], NotUsed]] = {
+    context.list[ListResource[T]].map { l =>
+      val initialSource = Source(l.items.map(l => WatchEvent(EventType.MODIFIED, l)))
+      val watchedSource = context.watchAllContinuously[T](Some(l.resourceVersion))
+
+      initialSource.concat(watchedSource)
+    }
+  }
+
+  def merge(implicit context: RequestContext,
+            sourceFormat: Format[S],
+            sourceListFormat: Format[ListResource[S]],
+            sourceResourceDefinition: ResourceDefinition[S],
+            targetFormat: Format[T],
+            targetListFormat: Format[ListResource[T]],
+            targetResourceDefinition: ResourceDefinition[T],
+            ec: ExecutionContext): akka.stream.scaladsl.Source[List[Event], NotUsed] = {
+
+    val sourceRaw = Source.fromFutureSource(watchSourceRaw)
+    val targetRaw = Source.fromFutureSource(watchTargetRaw)
+
+    val `WatchEvent[S]` = TypeCase[WatchEvent[S]]
+    val `WatchEvent[T]` = TypeCase[WatchEvent[T]]
+
+    Source
+      .combine(sourceRaw, targetRaw)(Merge(_))
+      .scan(Cache(List.empty, Map.empty)) { (acc, x) =>
+        x match {
+          case `WatchEvent[S]`(we) =>
+            val key = CacheEntryKey(name = we._object.name, namespace = we._object.namespace, kind = we._object.kind)
+            val targets = acc.cache.get(key).fold(List.empty[T])(_.ts)
+            val cache = acc.cache + (key -> CacheEntry(Some(we._object), targets))
+
+            we._type match {
+              case EventType.ADDED | EventType.MODIFIED => Cache(List(Modified(we._object, targets)), cache)
+              case EventType.DELETED                    => Cache(List(Deleted(we._object, targets)), cache)
```

Comment:
Should we not remove the source from the cache? This is a potential memory leak if we don't.